user/tests/wiring/watchdog/watchdog.cpp

```diff
@@ -55,8 +68,117 @@ test(WATCHDOG_02_default) {
     assertEqual(0, pushMailbox(MailboxEntry().type(MailboxEntry::Type::RESET_PENDING), 5000));
 }
 
-test(WATCHDOG_03_reset_reason) {
+test(WATCHDOG_02_default_2) {
+    assertTrue(System.resetReason() == RESET_REASON_WATCHDOG);
+    checkState(WatchdogState::DISABLED);
+}
+
+test(WATCHDOG_03_stopped_in_hibernate_mode_1) {
+    startWatchdog(WatchdogConfiguration().timeout(5s));
+    // Notify about a pending reset
+    assertEqual(0, pushMailbox(MailboxEntry().type(MailboxEntry::Type::RESET_PENDING), 10000));
+    System.sleep(SystemSleepConfiguration().mode(SystemSleepMode::HIBERNATE).gpio(WKP, FALLING));
+}
+
+test(WATCHDOG_03_stopped_in_hibernate_mode_2) {
+    assertTrue(System.resetReason() == RESET_REASON_POWER_MANAGEMENT);
+    checkState(WatchdogState::DISABLED);
+}
+
+test(WATCHDOG_04_continue_running_after_waking_up_from_none_hibernate_mode_1) {
+    startWatchdog(WatchdogConfiguration().timeout(5s).capabilities(WatchdogCap::RESET)); // SLEEP_RUNNING is cleared
+    assertEqual(0, pushMailbox(MailboxEntry().type(MailboxEntry::Type::RESET_PENDING), 1000));
+    System.sleep(SystemSleepConfiguration().mode(SystemSleepMode::STOP).duration(7s));
+    checkState(WatchdogState::STARTED);
+    // Notify about a pending reset
+    assertEqual(0, pushMailbox(MailboxEntry().type(MailboxEntry::Type::RESET_PENDING), 5000));
+}
+
+test(WATCHDOG_04_continue_running_after_waking_up_from_none_hibernate_mode_2) {
+    assertTrue(System.resetReason() == RESET_REASON_WATCHDOG);
+    checkState(WatchdogState::DISABLED);
+}
+
+test(WATCHDOG_05_system_reset_will_disable_watchdog_1) {
+    startWatchdog(WatchdogConfiguration().timeout(5s));
+    assertEqual(0, pushMailbox(MailboxEntry().type(MailboxEntry::Type::RESET_PENDING), 1000));
+    System.reset();
+}
+
+test(WATCHDOG_05_system_reset_will_disable_watchdog_2) {
+    assertTrue(System.resetReason() == RESET_REASON_USER);
+    checkState(WatchdogState::DISABLED);
+}
+
+#if HAL_PLATFORM_NRF52840
+
+test(WATCHDOG_06_default_running_in_none_hibernate_mode_1) {
+    startWatchdog(WatchdogConfiguration().timeout(5s));
+    // Notify about a pending reset
+    assertEqual(0, pushMailbox(MailboxEntry().type(MailboxEntry::Type::RESET_PENDING), 5000));
+    System.sleep(SystemSleepConfiguration().mode(SystemSleepMode::STOP).duration(10s));
+}
+
+test(WATCHDOG_06_default_running_in_none_hibernate_mode_2) {
+    assertTrue(System.resetReason() == RESET_REASON_WATCHDOG);
+    checkState(WatchdogState::DISABLED);
+}
+
+static retained uint32_t magick = 0;
```

Comment:
Are retained variables kept both on Gen 3 and P2/Tracker M after WDG reset?

Reply:
Not on P2 based platforms

Reply:
There is `hal_backup_ram_sync()`, so in theory you can perform that before WDG reset and have backup ram restored on boot.

Reply:
We stilll need to figure out how we can promise the `retained` key word works as expected implicitly.

Reply:
debug peripheral watchpoint for that memory region on writes?

Reply:
I'll create a SC ticket for tracking this problem.